File: interview_handwrite/bounded_blocking_queue.cpp
// Thread-safe bounded blocking queue (producer-consumer).
//
// Whiteboard talking points:
// - One mutex protects the deque.
// - Two condition variables: not_empty / not_full.
// - put() blocks when full; get() blocks when empty.
// - Always use while (not if) when waiting on a condition variable.
// - Amortized O(1) for put/get.

#include <cassert>
#include <chrono>
#include <condition_variable>
#include <deque>
#include <iostream>
#include <mutex>
#include <optional>
#include <stdexcept>
#include <thread>
#include <vector>

template <typename T>
class BoundedBlockingQueue {
 public:
    explicit BoundedBlockingQueue(size_t capacity) : capacity_(capacity) {
        if (capacity == 0) {
            throw std::invalid_argument("capacity must be positive");
        }
    }

    size_t size() const {
        std::lock_guard<std::mutex> lock(mutex_);
        return queue_.size();
    }

    void put(const T& item) {
        std::unique_lock<std::mutex> lock(mutex_);
        not_full_.wait(lock, [this] { return queue_.size() < capacity_; });
        queue_.push_back(item);
        not_empty_.notify_one();
    }

    bool put(const T& item, std::chrono::milliseconds timeout) {
        std::unique_lock<std::mutex> lock(mutex_);
        if (!not_full_.wait_for(lock, timeout,
                                [this] { return queue_.size() < capacity_; })) {
            return false;
        }
        queue_.push_back(item);
        not_empty_.notify_one();
        return true;
    }

    T get() {
        std::unique_lock<std::mutex> lock(mutex_);
        not_empty_.wait(lock, [this] { return !queue_.empty(); });
        T item = std::move(queue_.front());
        queue_.pop_front();
        not_full_.notify_one();
        return item;
    }

    std::optional<T> get(std::chrono::milliseconds timeout) {
        std::unique_lock<std::mutex> lock(mutex_);
        if (!not_empty_.wait_for(lock, timeout, [this] { return !queue_.empty(); })) {
            return std::nullopt;
        }
        T item = std::move(queue_.front());
        queue_.pop_front();
        not_full_.notify_one();
        return item;
    }

    bool try_put(const T& item) {
        std::lock_guard<std::mutex> lock(mutex_);
        if (queue_.size() >= capacity_) {
            return false;
        }
        queue_.push_back(item);
        not_empty_.notify_one();
        return true;
    }

    std::optional<T> try_get() {
        std::lock_guard<std::mutex> lock(mutex_);
        if (queue_.empty()) {
            return std::nullopt;
        }
        T item = std::move(queue_.front());
        queue_.pop_front();
        not_full_.notify_one();
        return item;
    }

 private:
    size_t capacity_;
    std::deque<T> queue_;
    mutable std::mutex mutex_;
    std::condition_variable not_empty_;
    std::condition_variable not_full_;
};

bool test_bounded_blocking_queue() {
    BoundedBlockingQueue<int> queue(2);
    std::vector<int> produced;
    std::vector<int> consumed;

    std::thread producer([&] {
        for (int i = 0; i < 5; ++i) {
            queue.put(i);
            produced.push_back(i);
        }
    });

    std::thread consumer([&] {
        for (int i = 0; i < 5; ++i) {
            consumed.push_back(queue.get());
        }
    });

    producer.join();
    consumer.join();

    for (int i = 0; i < 5; ++i) {
        assert(produced[i] == i);
        assert(consumed[i] == i);
    }
    return true;
}

int main() {
    assert(test_bounded_blocking_queue());
    std::cout << "bounded_blocking_queue: ok\n";
    return 0;
}

File: concurrency/bounded_blocking_queue.py
"""Thread-safe bounded blocking queue (producer-consumer)."""

from __future__ import annotations

import threading
import time
from collections import deque
from typing import Deque, Generic, List, Optional, TypeVar

T = TypeVar("T")


class BoundedBlockingQueue(Generic[T]):
    """
    Classic producer-consumer queue.

    - One mutex protects the deque.
    - Two condition variables: not_empty / not_full.
    - put() blocks when full; get() blocks when empty.

    get/put amortized O(1).
    """

    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._capacity = capacity
        self._queue: Deque[T] = deque()
        self._lock = threading.Lock()
        self._not_empty = threading.Condition(self._lock)
        self._not_full = threading.Condition(self._lock)

    def __len__(self) -> int:
        with self._lock:
            return len(self._queue)

    def put(self, item: T, timeout: Optional[float] = None) -> bool:
        with self._not_full:
            deadline = None if timeout is None else time.monotonic() + timeout
            while len(self._queue) >= self._capacity:
                if timeout is None:
                    self._not_full.wait()
                else:
                    remaining = deadline - time.monotonic()
                    if remaining <= 0:
                        return False
                    self._not_full.wait(remaining)
            self._queue.append(item)
            self._not_empty.notify()
            return True

    def get(self, timeout: Optional[float] = None) -> Optional[T]:
        with self._not_empty:
            deadline = None if timeout is None else time.monotonic() + timeout
            while not self._queue:
                if timeout is None:
                    self._not_empty.wait()
                else:
                    remaining = deadline - time.monotonic()
                    if remaining <= 0:
                        return None
                    self._not_empty.wait(remaining)
            item = self._queue.popleft()
            self._not_full.notify()
            return item

    def try_put(self, item: T) -> bool:
        with self._not_full:
            if len(self._queue) >= self._capacity:
                return False
            self._queue.append(item)
            self._not_empty.notify()
            return True

    def try_get(self) -> Optional[T]:
        with self._not_empty:
            if not self._queue:
                return None
            item = self._queue.popleft()
            self._not_full.notify()
            return item


def _test() -> None:
    q: BoundedBlockingQueue[int] = BoundedBlockingQueue(2)
    produced: List[int] = []
    consumed: List[int] = []

    def producer() -> None:
        for i in range(5):
            assert q.put(i)
            produced.append(i)

    def consumer() -> None:
        for _ in range(5):
            item = q.get()
            assert item is not None
            consumed.append(item)

    t1 = threading.Thread(target=producer)
    t2 = threading.Thread(target=consumer)
    t1.start()
    t2.start()
    t1.join()
    t2.join()
    assert produced == list(range(5))
    assert consumed == list(range(5))
    print("bounded_blocking_queue: ok")


if __name__ == "__main__":
    _test()
